p4-16/psa/PSA.mdk

```diff
@@ -970,80 +959,84 @@ The partial program below demonstrates how to clone a packet.
 
 To facilitate communication between the first pass and the second pass
-through the ingress pipeline, the resubmission mechanism supports
+through the ingress pipeline, the resubmission mechanism supportsr
 attaching optional metadata with the resubmitted packet. The metadata
 is generated during the first pass through the ingress pipeline, and
 used in the second pass. See Section [#sec-open-metadata] for the
 discussion.
 
 A PSA implementation provides a configuration bit `resubmit` to the
-PRE to enable the resubmission mechanism. If `resubmit` is set to
-true, the resubmission mechanism resends the original packet with the
-optional attached metadata. If `resubmit` is set to false, the
-resubmission mechanism is disabled and the original packet is not
-resubmitted even if the `add_metadata` method is invoked in the
-deparser.
+PRE to enable the resubmission mechanism. If true, the original packet
+is resubmitted with the optional resubmit metadata. If false, the
+resubmission mechanism is disabled and no assignments to
+`resubmit_meta` should be performed.
 
-```
-[INCLUDE=psa.p4:Resubmit_extern]
-
-```
 
 ## Packet Recirculation {#sec-recirculate}
 
-Packet recirculation sends the deparsed packet at the end of egress
-pipeline to the ingress parser. Unlike the previous two mechanisms,
-recirculation does not replicate packets. The modified packet itself
-is sent back to the ingress pipeline for further processing.
+Packet recirculation is a mechanism to repeat both ingress and egress
+processing on a packet.
 
-In PSA, recirculation is controlled by the `recirculate` bit in the
-egress pipeline. If the `recirculate` bit is set to true, the modified
-packet at the end of the pipeline is sent to a special recirculation
-port. The recirculation port overrides the value in `egress_port`.
+Packet recirculation happens at the end of the egress pipeline. When a
+packet is recirculated, the packet finishes both ingress and egress
+processing and re-enters the ingress parser after being deparsed by
+the egress deparser. The `ingress_port` of the recirculated packet is
+set to `PORT_RECIRCULATE`. The `packet_path` of the recirculated
+packet is changed to `RECIRCULATE`.
 
-```
-[INCLUDE=psa.p4:Recirculate_extern]
-
-```
-
-Users can leverage the existing `emit` methods in `packet_out` to
-attach additional metadata to the recirculated packet. Upon receiving
-the recirculated packet in ingress parser, the `packet_path` of the
-recirculated packet is set to `RECIRCULATE`. It is the user's
-responsibility to parse the recirculated packet correctly.
+Similar to packet resubmission, packet recirculation also supports
+attaching optional metadata with the recirculated packet. The metadata
+is generated during the first pass through the ingress and egress
+pipeline, and used in the second pass. See Section
+[#sec-open-metadata] for the discussion.
 
+The `recirculate` flag specifies whether a packet should be
```

Comment:
I can write an example to recriculate without the `recirculate` flag, if it works, then we can remove this paragraph in another PR?

Reply:
Sounds good.